rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -624,103 +690,53 @@ extern "C" rmw_ret_t rmw_client_set_on_new_response_callback(
 {
+  CddsSubscription * sub_event;
+  CddsPublisher * pub_event;
+  user_callback_data_t * data;
   RMW_CHECK_ARGUMENT_FOR_NULL(rmw_event, RMW_RET_INVALID_ARGUMENT);
+
   switch (rmw_event->event_type) {
     case RMW_EVENT_LIVELINESS_CHANGED:
-      {
-        auto sub_event = static_cast<CddsSubscription *>(rmw_event->data);
-        event_set_callback(
-          sub_event, DDS_LIVELINESS_CHANGED_STATUS_ID,
-          callback, user_data);
-        break;
-      }
-
     case RMW_EVENT_REQUESTED_DEADLINE_MISSED:
-      {
-        auto sub_event = static_cast<CddsSubscription *>(rmw_event->data);
-        event_set_callback(
-          sub_event, DDS_REQUESTED_DEADLINE_MISSED_STATUS_ID,
-          callback, user_data);
-        break;
-      }
-
     case RMW_EVENT_REQUESTED_QOS_INCOMPATIBLE:
-      {
-        auto sub_event = static_cast<CddsSubscription *>(rmw_event->data);
-        event_set_callback(
-          sub_event, DDS_REQUESTED_INCOMPATIBLE_QOS_STATUS_ID,
-          callback, user_data);
-        break;
-      }
-
     case RMW_EVENT_MESSAGE_LOST:
+    case RMW_EVENT_SUBSCRIPTION_MATCHED:
+    case RMW_EVENT_SUBSCRIPTION_UNMATCHED:
       {
-        auto sub_event = static_cast<CddsSubscription *>(rmw_event->data);
-        event_set_callback(
-          sub_event, DDS_SAMPLE_LOST_STATUS_ID,
-          callback, user_data);
+        sub_event = static_cast<CddsSubscription *>(rmw_event->data);
+        data = &sub_event->user_callback_data;
         break;
       }
-
     case RMW_EVENT_LIVELINESS_LOST:
-      {
-        auto pub_event = static_cast<CddsPublisher *>(rmw_event->data);
-        event_set_callback(
-          pub_event, DDS_LIVELINESS_LOST_STATUS_ID,
-          callback, user_data);
-        break;
-      }
-
     case RMW_EVENT_OFFERED_DEADLINE_MISSED:
-      {
-        auto pub_event = static_cast<CddsPublisher *>(rmw_event->data);
-        event_set_callback(
-          pub_event, DDS_OFFERED_DEADLINE_MISSED_STATUS_ID,
-          callback, user_data);
-        break;
-      }
-
     case RMW_EVENT_OFFERED_QOS_INCOMPATIBLE:
+    case RMW_EVENT_PUBLICATION_MATCHED:
+    case RMW_EVENT_PUBLICATION_UNMATCHED:
       {
-        auto pub_event = static_cast<CddsPublisher *>(rmw_event->data);
-        event_set_callback(
-          pub_event, DDS_OFFERED_INCOMPATIBLE_QOS_STATUS_ID,
-          callback, user_data);
+        pub_event = static_cast<CddsPublisher *>(rmw_event->data);
+        data = &pub_event->user_callback_data;
         break;
       }
-
     case RMW_EVENT_INVALID:
       {
         return RMW_RET_INVALID_ARGUMENT;
       }
   }
+
+  // Set the user callback data
+  data->event_callback[rmw_event->event_type] = callback;
+  data->event_data[rmw_event->event_type] = user_data;
+
+  if (callback && data->event_unread_count[rmw_event->event_type]) {
+    // Push events happened before having assigned a callback
+    callback(user_data, data->event_unread_count[rmw_event->event_type]);
```

Comment:
This will notify the user application `matched/unmatched` event that happened before? I am not sure if this is what user expects? probably user application is not interested in the events in the past for this case?

at least, we would want to have the same behavior for rmw implementation.

Reply:
Yes. 
But for fastrtps, only matched/unmatched event don't notify previoius status. Other event will do.  
So for keeping the same behavior, we have to deal with matched/unmatched individually.

Reply:
Need to add below codes only for matched/unmatched event

```c++
  switch (rmw_event->event_type) {
    case RMW_EVENT_SUBSCRIPTION_MATCHED:
      data->event_unread_count[RMW_EVENT_SUBSCRIPTION_MATCHED] = 0;
      break;
    case RMW_EVENT_SUBSCRIPTION_UNMATCHED:
      data->event_unread_count[RMW_EVENT_SUBSCRIPTION_UNMATCHED] = 0;
      break;
    case RMW_EVENT_PUBLICATION_MATCHED:
      data->event_unread_count[RMW_EVENT_PUBLICATION_MATCHED] = 0;
      break;
    case RMW_EVENT_PUBLICATION_UNMATCHED:
      data->event_unread_count[RMW_EVENT_PUBLICATION_UNMATCHED] = 0;
      break;
    default:
      break;
  }
```